crates/telio-network-monitors/src/monitor.rs

```diff
@@ -0,0 +1,127 @@
+    windows_monitor_handle: crate::windows::SafeHandle,
+}
+
+impl NetworkMonitor {
+    /// Sets up and spawns network monitor
+    pub async fn new<G>(if_addr: G) -> Result<NetworkMonitor, GetIFError>
+    where
+        G: GetIfAddrs + Clone,
+    {
+        if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+            *guard = local_interfaces::gather_local_interfaces(&if_addr)?;
+            telio_log_debug!("local cache {:?}", *guard);
+        }
+
+        let nw_path_monitor_monitor_handle = Some(tokio::spawn({
+            let mut notify = PATH_CHANGE_BROADCAST.subscribe();
+            async move {
+                loop {
+                    match notify.recv().await {
+                        Ok(()) => match local_interfaces::gather_local_interfaces(&if_addr) {
+                            Ok(v) => {
+                                if let Ok(mut guard) = LOCAL_ADDRS_CACHE.lock() {
+                                    telio_log_debug!("Updating local addr cache");
+                                    *guard = v;
+                                }
+                            }
+                            Err(e) => telio_log_warn!("Unable to get local interfaces {e}"),
+                        },
+                        Err(e) => {
+                            telio_log_warn!("Failed to receive new path for sockets ({e}): ");
+                        }
+                    }
+                }
+            }
+        }));
+
+        #[cfg(any(target_os = "macos", target_os = "ios", target_os = "tvos"))]
+        NETWORK_PATH_MONITOR_START.call_once(crate::mac::setup_network_monitor);
+
+        #[cfg(target_os = "linux")]
+        let linux_monitor_handle = crate::linux::setup_network_monitor().await;
+
+        #[cfg(target_os = "windows")]
+        let windows_monitor_handle = crate::windows::setup_network_monitor();
+
+        Ok(Self {
+            nw_path_monitor_monitor_handle,
+            #[cfg(target_os = "linux")]
+            linux_monitor_handle,
+            #[cfg(target_os = "windows")]
+            windows_monitor_handle,
+        })
+    }
+}
+
+impl Drop for NetworkMonitor {
+    fn drop(&mut self) {
+        if let Some(handle) = &self.nw_path_monitor_monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "linux")]
+        if let Some(handle) = &self.linux_monitor_handle {
+            handle.abort();
+        }
+        #[cfg(target_os = "windows")]
+        if !self.windows_monitor_handle.0.is_null() {
+            crate::windows::deregister_network_monitor(self.windows_monitor_handle.clone());
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use telio_utils::local_interfaces::MockGetIfAddrs;
+
+    use super::*;
+    use std::net::Ipv4Addr;
+
+    #[tokio::test]
+    async fn test_path_change_notification() {
+        let mut get_if_addrs_mock = MockGetIfAddrs::new();
+        get_if_addrs_mock.expect_get().times(2).returning(|| {
+            Ok(vec![if_addrs::Interface {
```

Comment:
something's off, this test is passing even though this expectation is not met. I was playing with it locally and changed time(2) -> times(1) and although the logs print:
`MockGetIfAddrs::get: Expectation(<anything>) called 2 times which is more than the expected 1`
the test still passes, which shouldn't right?

Reply:
you're probably missing the checkpoint() at the end of test.

Reply:
.. or mock is moving out of scope and thus cannot assert the expectations.